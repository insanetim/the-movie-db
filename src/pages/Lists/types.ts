import { ErrorMsg, HandlePagination } from 'src/interfaces/global.interface'
import { IListsList } from 'src/interfaces/list.interface'

export type ListsHookReturn = {
  error: ErrorMsg
  handleCreateList: () => () => void
  handlePagination: HandlePagination
  lists: IListsList | null
  loading: boolean
}
