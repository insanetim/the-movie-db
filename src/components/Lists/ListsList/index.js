import ListsList from './component'

export default ListsList
