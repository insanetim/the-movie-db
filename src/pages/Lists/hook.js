import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { not, isEmpty } from 'ramda'

import { showModal } from 'src/state/app/actions'
import { fetchLists } from 'src/state/lists/actions'
import { accountSelector } from 'src/state/session/selectors'
import { listsSelector, listsPageSelector, listsLoadingSelector, listsErrorSelector } from 'src/state/lists/selectors'

const useContainer = () => {
  const dispatch = useDispatch()
  const account = useSelector(accountSelector)
  const lists = useSelector(listsSelector)
  const page = useSelector(listsPageSelector)
  const loading = useSelector(listsLoadingSelector)
  const error = useSelector(listsErrorSelector)

  const handleClick = () => {
    dispatch(
      showModal({
        modalType: 'CREATE_LIST_MODAL'
      })
    )
  }

  useEffect(() => {
    if (not(isEmpty(account))) {
      dispatch(fetchLists(page))
    }
  }, [account, page])

  return { lists, loading, error, handleClick }
}

export default useContainer
