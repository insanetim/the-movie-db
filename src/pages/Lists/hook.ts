import { isNotNil } from 'ramda'
import { useEffect } from 'react'
import { useSearchParams } from 'react-router-dom'
import { useAppDispatch, useAppSelector } from 'src/hooks/useRedux'
import { showModal } from 'src/store/app/actions'
import { accountSelector } from 'src/store/auth/selectors'
import { fetchLists } from 'src/store/lists/actions'
import {
  listsErrorSelector,
  listsLoadingSelector,
  listsSelector,
} from 'src/store/lists/selectors'
import getParams from 'src/utils/helpers/getParams'

import { ListsHook } from './types'

const useContainer = (): ListsHook => {
  const dispatch = useAppDispatch()
  const account = useAppSelector(accountSelector)
  const lists = useAppSelector(listsSelector)
  const loading = useAppSelector(listsLoadingSelector)
  const error = useAppSelector(listsErrorSelector)
  const [searchParams, setSearchParams] = useSearchParams()
  const page = searchParams.get('page') ?? '1'

  const handlePagination = (page: number) => {
    setSearchParams(getParams({ page }))
  }

  const handleCreateList = () => {
    const onSuccess = () => {
      if (page === '1') {
        dispatch(fetchLists(page))
      } else {
        setSearchParams({})
      }
    }

    dispatch(
      showModal({
        modalProps: { onSuccess },
        modalType: 'MODAL_CREATE_LIST',
      })
    )

    return onSuccess
  }

  useEffect(() => {
    if (isNotNil(account)) {
      dispatch(fetchLists(page))
    }
  }, [account, page, dispatch])

  return { error, handleCreateList, handlePagination, lists, loading }
}

export default useContainer
