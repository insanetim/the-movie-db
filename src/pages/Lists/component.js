import { PlusCircleOutlined } from '@ant-design/icons'

import ListsList from 'src/components/Lists/ListsList'
import Loading from 'src/components/UI/Loading'
import Error from 'src/components/UI/Error'
import PageTitle from 'src/components/UI/PageTitle'
import useContainer from './hook'

const Lists = () => {
  const { lists, loading, error, handleClick } = useContainer()

  return (
    <div className='container'>
      <PageTitle>
        My Lists <PlusCircleOutlined onClick={handleClick} />
      </PageTitle>
      {loading && (
        <div className='top-margin'>
          <Loading />
        </div>
      )}
      {error && <Error error={error} />}
      {!loading && !error && <ListsList lists={lists} />}
    </div>
  )
}

export default Lists
