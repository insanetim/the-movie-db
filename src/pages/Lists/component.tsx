import { PlusCircleOutlined } from '@ant-design/icons'
import { isEmpty, isNotNil } from 'ramda'
import { Helmet } from 'react-helmet'
import ListsList from 'src/components/Lists/ListsList'
import Empty from 'src/components/UI/Empty/component'
import Error from 'src/components/UI/Error'
import Loading from 'src/components/UI/Loading'
import PageTitle from 'src/components/UI/PageTitle'
import Pagination from 'src/components/UI/Pagination'
import metaTitle from 'src/utils/helpers/metaTitle'

import useContainer from './hook'

const Lists: React.FC = () => {
  const { error, handleCreateList, handlePagination, lists, loading } =
    useContainer()

  let content = <Empty />

  if (loading) {
    content = (
      <div className='top-margin'>
        <Loading />
      </div>
    )
  }
  if (error) {
    content = <Error error={error} />
  }
  if (isNotNil(lists) && !isEmpty(lists.results)) {
    content = (
      <>
        <ListsList lists={lists.results} />
        {lists.total_pages > 1 && (
          <Pagination
            current={lists.page}
            onChange={handlePagination}
            pageSize={20}
            total={lists.total_results}
          />
        )}
      </>
    )
  }

  return (
    <>
      <Helmet title={metaTitle('My Lists')} />
      <div className='container'>
        <PageTitle>
          My Lists{' '}
          <PlusCircleOutlined
            data-testid='createListBtn'
            onClick={handleCreateList}
          />
        </PageTitle>
        {content}
      </div>
    </>
  )
}

export default Lists
