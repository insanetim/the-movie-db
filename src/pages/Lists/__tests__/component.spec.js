import { render } from '@testing-library/react'

import Lists from '../component'
import Wrapper from '../../../__mocks__/wrapperMock'

const mockedHookData = {
  lists: {
    results: [
      {
        id: 1,
        title: 'test/title',
        overview: 'test/overview',
        poster_path: 'test/image'
      }
    ]
  },
  loading: false,
  error: null,
  handleClick: jest.fn()
}
jest.mock('../hook', () => jest.fn(() => mockedHookData))

describe('Lists component', () => {
  afterAll(() => {
    jest.unmock('../hook')
  })

  it('matches snapshot', () => {
    const { asFragment } = render(<Lists />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })

  it('matches snapshot with loading', () => {
    mockedHookData.loading = true
    const { asFragment } = render(<Lists />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })

  it('matches snapshot with error', () => {
    mockedHookData.loading = false
    mockedHookData.error = { message: 'test/error' }
    const { asFragment } = render(<Lists />, { wrapper: Wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
})
