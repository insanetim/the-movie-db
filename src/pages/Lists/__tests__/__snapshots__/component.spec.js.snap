// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Fragment>
  <Row
    justify="center"
  >
    <Col
      span={20}
    >
      <div
        className="top-margin"
      >
        <ForwardRef>
          My Lists 
          <ForwardRef(PlusCircleOutlined)
            onClick={[MockFunction]}
          />
        </ForwardRef>
      </div>
    </Col>
  </Row>
  <ListsList
    lists={
      {
        "results": [
          {
            "id": 1,
            "overview": "test/overview",
            "poster_path": "test/image",
            "title": "test/title",
          },
        ],
      }
    }
  />
</Fragment>
`;

exports[`matches snapshot with loading 1`] = `
<Fragment>
  <Row
    justify="center"
  >
    <Col
      span={20}
    >
      <div
        className="top-margin"
      >
        <ForwardRef>
          My Lists 
          <ForwardRef(PlusCircleOutlined)
            onClick={[MockFunction]}
          />
        </ForwardRef>
      </div>
    </Col>
  </Row>
  <ListsList
    lists={{}}
  />
</Fragment>
`;
